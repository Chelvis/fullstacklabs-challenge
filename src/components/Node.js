import {
  Accordion, AccordionDetails, AccordionSummary, Box, makeStyles, Typography
} from "@material-ui/core";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import PropTypes from "prop-types";
import React from "react";
import colors from "../constants/colors";
import Status from "./Status";

const Node = ({ node, expanded, toggleNodeExpanded }) => {
  const classes = useStyles();

  const buildBlocks = (blocks) => {
    if (blocks === undefined) {
      return <Typography>Impossible to retreive data.</Typography>
    } else if (!blocks.length) {
      return <Typography>No data to show.</Typography>
    }
    return blocks.map((block, index) => {
      return <Box key={index} className={classes.block}>
        <Typography className={classes.block_id}>{block.id.padStart(3, '0')}</Typography>
        <Typography>{block.attributes.data}</Typography>
      </Box>
    })
  }
  
  return (
    <Accordion
      elevation={3}
      className={classes.root}
      expanded={expanded}
      onChange={() => toggleNodeExpanded(node)}
    >
      <AccordionSummary
        className={classes.summary}
        classes={{
          expandIcon: classes.icon,
          content: classes.content,
          expanded: classes.expanded,
        }}
        expandIcon={<ExpandMoreIcon />}
      >
        <Box className={classes.summaryContent}>
          <Box>
            <Typography variant="h5" className={classes.heading}>
              {node.name || "Unknown"}
            </Typography>
            <Typography
              variant="subtitle1"
              className={classes.secondaryHeading}
            >
              {node.url}
            </Typography>
          </Box>
          <Status loading={node.loading} online={node.online} />
        </Box>
      </AccordionSummary>
      <AccordionDetails>
        <Box className={classes.block_wrapper}>
          {buildBlocks(node.blocks)}
        </Box>
      </AccordionDetails>
    </Accordion>
  );
};

const useStyles = makeStyles((theme) => ({
  root: {
    margin: "16px 0",
    boxShadow: "0px 3px 6px 1px rgba(0,0,0,0.15)",
    "&:before": {
      backgroundColor: "unset",
    },
  },
  summary: {
    padding: "0 24px",
  },
  summaryContent: {
    display: "flex",
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    width: "100%",
    paddingRight: 20,
  },
  icon: {
    color: colors.faded,
  },
  content: {
    margin: "10px 0 !important", // Avoid change of sizing on expanded
  },
  expanded: {
    "& $icon": {
      paddingLeft: 0,
      paddingRight: 12,
      top: -10,
      marginRight: 0,
    },
  },
  heading: {
    fontSize: theme.typography.pxToRem(17),
    display: "block",
    color: colors.text,
    lineHeight: 1.5,
  },
  secondaryHeading: {
    fontSize: theme.typography.pxToRem(14),
    color: colors.faded,
    lineHeight: 2,
  },
  block_wrapper: {
    width: '100%'
  },
  block: {
    background: 'rgba(0,0,0,.12)',
    padding: '8px',
    borderRadius: '2px',
    margin: '4px 0'
  },
  block_id: {
    color: '#304FFE',
    fontSize: '10px'
  }
}));

Node.propTypes = {
  node: PropTypes.shape({
    url: PropTypes.string,
    online: PropTypes.bool,
    name: PropTypes.string,
    loading: PropTypes.bool,
  }).isRequired,
  expanded: PropTypes.bool,
  toggleNodeExpanded: PropTypes.func.isRequired,
};

export default Node;
