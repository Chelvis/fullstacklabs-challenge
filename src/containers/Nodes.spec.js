import { shallow } from "enzyme";
import React from "react";
import { Provider } from "react-redux";
import { create } from "react-test-renderer";
import configureMockStore from "redux-mock-store";
import thunk from 'redux-thunk';
import Node from "../components/Node";
import ConnectedNodes, { Nodes } from "./Nodes";

describe("<Nodes />", () => {
  const actions = {
    checkNodeStatuses: jest.fn(),
    getNodesBlocks: jest.fn()
  };

  const nodes = {
    list: [
      {
        url: 'https://thawing-springs-53971.herokuapp.com',
        online: false,
        name: 'Node 1',
        loading: false,
        blocks: [{id: '1', attributes: {data: 'test 1'}}]
      },
      {
        url: 'https://secret-lowlands-62331.herokuapp.com',
        online: false,
        name: 'Node 2',
        loading: false,
        blocks: [{id: '2', attributes: {data: 'test 2'}}]
      }
    ]
  };

  it("should contain <Node />", () => {
    const wrapper = shallow(
      <Nodes
        actions={actions}
        nodes={nodes}
      />
    );

    expect(wrapper.find(Node).length).toEqual(2);
  });

  it("should match snapshot", () => {
    const middlewares = [thunk];
    const store = configureMockStore(middlewares)({nodes});
    const component = create(
      <Provider store={store}>
        <ConnectedNodes />
      </Provider>
    );
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
