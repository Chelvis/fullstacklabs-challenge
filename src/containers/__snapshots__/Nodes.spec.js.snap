// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nodes /> should match snapshot 1`] = `
<div
  className="MuiBox-root MuiBox-root-1"
>
  <h1
    className="MuiTypography-root MuiTypography-h4"
  >
    <strong
      style={
        Object {
          "color": "#000",
        }
      }
    >
      Nodes
    </strong>
  </h1>
  <div
    className="MuiPaper-root MuiAccordion-root makeStyles-root-2 MuiAccordion-rounded MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      aria-disabled={false}
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root makeStyles-summary-3"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content makeStyles-content-6"
      >
        <div
          className="MuiBox-root MuiBox-root-13 makeStyles-summaryContent-4"
        >
          <div
            className="MuiBox-root MuiBox-root-14"
          >
            <h5
              className="MuiTypography-root makeStyles-heading-8 MuiTypography-h5"
            >
              Node 1
            </h5>
            <h6
              className="MuiTypography-root makeStyles-secondaryHeading-9 MuiTypography-subtitle1"
            >
              https://thawing-springs-53971.herokuapp.com
            </h6>
          </div>
          <div
            className="MuiBox-root MuiBox-root-19"
          >
            <span
              className="makeStyles-dot-15 makeStyles-dot-17"
            />
            <span
              className="makeStyles-text-16 makeStyles-text-18"
            >
              OFFLINE
            </span>
          </div>
        </div>
      </div>
      <div
        aria-disabled={false}
        aria-hidden={true}
        className="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon makeStyles-icon-5 MuiIconButton-edgeEnd"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={null}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              className="MuiAccordionDetails-root"
            >
              <div
                className="MuiBox-root MuiBox-root-20 makeStyles-block_wrapper-10"
              >
                <div
                  className="MuiBox-root MuiBox-root-21 makeStyles-block-11"
                >
                  <p
                    className="MuiTypography-root makeStyles-block_id-12 MuiTypography-body1"
                  >
                    001
                  </p>
                  <p
                    className="MuiTypography-root MuiTypography-body1"
                  >
                    test 1
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiPaper-root MuiAccordion-root makeStyles-root-2 MuiAccordion-rounded MuiPaper-elevation3 MuiPaper-rounded"
  >
    <div
      aria-disabled={false}
      aria-expanded={false}
      className="MuiButtonBase-root MuiAccordionSummary-root makeStyles-summary-3"
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role="button"
      tabIndex={0}
    >
      <div
        className="MuiAccordionSummary-content makeStyles-content-6"
      >
        <div
          className="MuiBox-root MuiBox-root-22 makeStyles-summaryContent-4"
        >
          <div
            className="MuiBox-root MuiBox-root-23"
          >
            <h5
              className="MuiTypography-root makeStyles-heading-8 MuiTypography-h5"
            >
              Node 2
            </h5>
            <h6
              className="MuiTypography-root makeStyles-secondaryHeading-9 MuiTypography-subtitle1"
            >
              https://secret-lowlands-62331.herokuapp.com
            </h6>
          </div>
          <div
            className="MuiBox-root MuiBox-root-26"
          >
            <span
              className="makeStyles-dot-15 makeStyles-dot-24"
            />
            <span
              className="makeStyles-text-16 makeStyles-text-25"
            >
              OFFLINE
            </span>
          </div>
        </div>
      </div>
      <div
        aria-disabled={false}
        aria-hidden={true}
        className="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon makeStyles-icon-5 MuiIconButton-edgeEnd"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role={null}
        tabIndex={null}
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      className="MuiCollapse-container MuiCollapse-hidden"
      style={
        Object {
          "minHeight": "0px",
        }
      }
    >
      <div
        className="MuiCollapse-wrapper"
      >
        <div
          className="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              className="MuiAccordionDetails-root"
            >
              <div
                className="MuiBox-root MuiBox-root-27 makeStyles-block_wrapper-10"
              >
                <div
                  className="MuiBox-root MuiBox-root-28 makeStyles-block-11"
                >
                  <p
                    className="MuiTypography-root makeStyles-block_id-12 MuiTypography-body1"
                  >
                    002
                  </p>
                  <p
                    className="MuiTypography-root MuiTypography-body1"
                  >
                    test 2
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
